### PR TITLE
Various Documentation improvements

### PR DESCRIPTION
**Pull request by jaxoncreed**
Covers:
 - Describes the use case and implementation of PoP tokens
 - The description of asking for consent on the example page is more in line with reality
 - Extremely detailed walkthrough

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -99,6 +99,8 @@ WebID-OIDC makes the following changes to the base OpenID Connect protocol
 * Specifies the [Authorized OIDC Issuer
   Discovery](#authorized-oidc-issuer-discovery) process (used as part of
   Provider Confirmation, and during Provider Selection steps).
+* Utilizes [PoP tokens](https://tools.ietf.org/html/rfc7800) to securely
+  query a wide array of resource providers.
```
Comment (michielbdejong) [inline README.md]: ".. to let the client access a wide .."

File: README.md [hunk 2]
```diff
@@ -300,6 +302,24 @@ that profile, she would add the following triple to her profile:
...
+
+Unlike standard implementations of OIDC, WebID-OIDC must deal with a number of RPs many of which the OP will not know about. OIDC defines the `aud` claim which defines the RPs for which a token can be used.
+
+However, given Solid's use case, a token should be usable for any RP so the user may federate a query across multiple Pods, so the `aud`ience cannot be constrained. Yet, an unconstrained `aud`ience opens up the possibility of token stealing. In this case, a user sends a request to `evilPod.example`. The Pod returns the requested information, but now has the user's token and may pretend to be the user on any other Pod in the world.
+
+#### The Solution
+
+The solution employs [Proof of Possession (PoP) tokens](https://tools.ietf.org/html/rfc7800) changing the way the Bearer token is constructed:
+
+ 1. A client application generates a public and private key.
```
Comment (jaxoncreed) [inline README.md]: Mention that these keys are short lived

File: README.md [hunk 3]
```diff
@@ -300,6 +302,24 @@ that profile, she would add the following triple to her profile:
...
+ 2. The client generates a request `JWT` just as it would under normal OIDC with the addition of a `key` field containing the public key.
+ 3. Authentication proceeds normally and yeilds a signed `id_token` where the `aud`ience is the client application (represented by the `origin` of the provided `redirect_uri`) and an additional field `cnf` is provided containing the client's public key.
```
Comment (michielbdejong) [inline README.md]: yeilds -> yields

**Review by michielbdejong — COMMENTED** [review]
in the png, '6. Generates and id_token' -> '6. Generates an id_token'
step 5 of 'sending request' seems incorrect to me, IIUC RP only contacts OP to retrieve their jwks, not to ask them to perform validation.

**Comment by zenomt** [comment]
is this PR just to document how the reference implementation currently works, or to also suggest new functionality that is yet to be implemented?  i see there's mention (in "README.md -> The Solution, step 3"), of the origin of the redirect_uri becoming "the `aud`ience".

i haven't checked the source to see if that's what the current NSS OP actually does, but regardless that has some issues (most importantly, the `aud` MUST include the client_id, which as discussed in #12 the client_id can't be just the redirect_uri origin, or even the redirect_uri).  i think at least that part should be called out as "as currently implemented" or "proposed behavior" so that it can be commented on appropriately.

if this is documenting what's currently implemented, or at least what's currently designed/planned, then discussing issues with it isn't appropriate in this PR's conversation, and should be in an actual Issue once this change is merged. however, if this PR is also proposing new behavior, then discussing issues with the new behavior would be appropriate here.

**Comment by michielbdejong** [comment]
@jaxoncreed ping

**Comment by jaxoncreed** [comment]
I've now finished the detailed application workflow, so I think we're good to go. It's not the 100% rewrite this needs, but it provides some much better context.

**Review by michielbdejong — DISMISSED** [review]
I didn't read/verify all of it but looks like a great improvement over the current text. We can always iterate if we discover errata. A second review from @zenomt and/or @dmitrizagidulin would be helpful!

**Comment by michielbdejong** [comment]
We should probably update the list of people with write access to this repo to match the list of people who know most about how webid-oidc works. Created https://github.com/solid/webid-oidc-spec/issues/32 in relation to this.

**Comment by michielbdejong** [comment]
@kjetilk @Mitzi-Laszlo @timbl @justinwb @RubenVerborgh: two of you need to approve this before it can be merged.

**Comment by RubenVerborgh** [comment]
@michielbdejong Will do; useful to use the "assign reviewers" functionality for that.

File: README.md [hunk 4]
```diff
@@ -43,7 +43,7 @@ See also: [Motivation for WebID-OIDC](motivation.md).
 ### Benefits and Capabilities
 
 * Fully decentralized cross-domain authentication (any peer node can serve as
-  an identity provider as well as a relying party to any other node)
+  an identity provider as well as a relying party to any other node) made possible by [PoP Tokens](https://tools.ietf.org/html/rfc7800).
```
Comment (RubenVerborgh) [inline README.md]: Perhaps say a word about what they are, or at least expand acronym?

File: README.md [hunk 5]
```diff
@@ -300,6 +302,24 @@ that profile, she would add the following triple to her profile:
 <#me> solid:oidcIssuer <https://provider.com> .
 ```
 
+## Securing tokens for multiple resource servers
+
+#### The Problem
+
+Unlike standard implementations of OIDC, WebID-OIDC must deal with a number of RSs many of which the OP will not know about. OIDC defines the `aud` claim which defines the RSs for which a token can be used.
```
Comment (RubenVerborgh) [inline README.md]: But we are using a standard, right? Important that people understand this is not a custom extension.

File: application-workflow-detailed.md [hunk 6]
```diff
@@ -0,0 +1,442 @@
...
+```
+Data:
+```json
+{
+  grant_types: ["implicit"]
+  issuer: "https://secureauth.example"
+  redirect_uris: ["https://www.decentphotos.example/"]
+  response_types: ["id_token token"]
+  scope: "openid profile"
+}
```
Comment (RubenVerborgh) [inline application-workflow-detailed.md]: invalid JSON

**Review by RubenVerborgh — DISMISSED** [review]
Technically correct, but I feel that we are too much describing an existing spec here. We should point more and explain less (unless the actual specs are not clear). However, good from a documentation perspective, but need to revisited when turning into an actual spec, where we want to avoid repetition of other specs.

**Comment by zenomt** [comment]
in the current version of the sequence diagram, authorization steps 3 and 4 (request OP configuration) are to Alice's POD, but i think you mean for them to go to her OP.

File: application-workflow-detailed.md [hunk 7]
```diff
@@ -0,0 +1,442 @@
...
+
+This document outlines, in detail, the login and request process for an application using WebId-OIDC. In general, our user, *Alice* will be using a thrid-party application at `https://www.decentphotos.example`  to access data on both her pod at `https://alice.example` and her friend, Bob's pod at `https://bob.example`.
+
+## Actors
+
+In this example a multitude of actors are at play:
+
+**Alice** - Alice will be providing consent for decentphotos to use her pod. Let's assume that Alice is using a standard web browser.
+
+**Bob's Pod (RS)** - We will be trying to access photos Bob's Pod, known in the OIDC world as a Resource Server (RS). Bob is a friend of Alice. For this use case, let's assume that Bob has previously indicated via access control that Alice may access his photo using any app. You can read more about access control [here](https://github.com/solid/web-access-control-spec#referring-to-origins-ie-web-apps). For this example, bob's pod is at `bob.solid.example`.
```
Comment (justinwb) [inline application-workflow-detailed.md]: - typo - should be "trying to access photos stored in Bob's Pod"
- need to delineate between app and webapp. the link is to add trusted origins, but it says that bob indicated through access control that alice could access his photo using any "app".

File: application-workflow-detailed.md [hunk 8]
```diff
@@ -0,0 +1,442 @@
...
+
+**Alice's OP** - Alice's OpenID Provider (OP), also known as an IDP (Identity Provider), is the service responsible for authorizing our thrid-party app by providing it with the tokens necessary to gain access to any pod. In this demo, alice's OP is at `secureauth.example`.
```
Comment (justinwb) [inline application-workflow-detailed.md]: typo - thrid -> third

File: application-workflow-detailed.md [hunk 9]
```diff
@@ -0,0 +1,442 @@
...
+
+## Application Flow
+
+![Application Login Flow](ApplicationWorkflow.png)
+
+### Authorization
+
+Before any requests can be made, Alice must log in:
+
+#### 1. Alice naviages to www.decentphotos.example
```
Comment (justinwb) [inline application-workflow-detailed.md]: typo: naviages -> navigates

File: application-workflow-detailed.md [hunk 10]
```diff
@@ -0,0 +1,442 @@
...
+
+Alice has heard of a great new site that allows her to view her friend's photos and tag faces. She navigates to `www.decentphotos.example` via he web browser which returns and html page. This page contains JavaScript that will help with the authentication process.
```
Comment (justinwb) [inline application-workflow-detailed.md]: typo: via he -> via her
typo: returns and -> returns an

File: application-workflow-detailed.md [hunk 11]
```diff
@@ -0,0 +1,442 @@
...
+
+#### 2. Alice clicks the "Connect" button
+
+Before decentphotos can start displaying images, Alice needs to start the process of providing consent. To do so, she must either provider her webId (`https://alice.coolpod.example/profile/card#me`) or the service the url of her OP (`https://secureauth.example`)
```
Comment (justinwb) [inline application-workflow-detailed.md]: typo: either provider -> either provide
typo: or the service the url -> or the url

File: application-workflow-detailed.md [hunk 12]
```diff
@@ -0,0 +1,442 @@
...
+  "scope": "openid profile"
+}
+```
+
+Each of these communicates something about the new client to the OP:
+ - `grant_types`: A list of [OIDC grant types](http://docs.identityserver.io/en/latest/topics/grant_types.html) this client will use. `implicit` is great for web applications.
+ - `issuer`: Alice's OP
+ - `redirect_uris`: Redirect uris provided at the client registration stage state which redirect uris are valid during the authorization stage.
+ - `response_types`: A list of [OIDC response types](https://openid.net/specs/oauth-v2-multiple-response-types-1_0.html) the client can use. `id_token token` means that it should return both an Id Token (needed to identify this application later) and an access token.
+ - `scope`: OIDC uses scope as a way of defining what a client can have acces to. However, Solid has it's own access control system, so scope will always be `openid profile`
```
Comment (justinwb) [inline application-workflow-detailed.md]: typo: can have acces to -> can have access to
typo: Solid has it's -> Solid has its

File: application-workflow-detailed.md [hunk 13]
```diff
@@ -0,0 +1,442 @@
...
+    "n": "2zymE1Pt9ONnlijH0yP7b-fU3gEloybOqMmYJ0sStY1EMCo8TMtKfFm9r0RuzrTK02N_uC66EpGShOVx2i2DH1mekVCyewkZI3pCdedQY9t8lcEGPNoqEpH5bFCYObnwt-luTkD87hE__LmkkA_p4o90ViDE0pIhalMikyIK16fjWr8vaZytjmEiQ-8gy5K6YIJTtrHIr8c7vuzks2Py9WoLe8o2U0ohamEgvB8alRAPQlRWbPxwdnkVo9fr9PPPzX3txxkd56swjwOdRVjVpfehQUtt0nLu040cWpvkX0SUx6vLNIrbQ1rS9Qy12GQxijd8_R7u0ZHWPhFJwchcjw"
+  }
+}
+ ```
+
+ Notice that the `key` contains the public key generated earlier.
+
+#### 13. Gets Alice's Consent
+
+Given everything checks out in the last step, the OP should redirect to it's login screen. The actual implementation of this is completely up to the OP. A user can log in with her password, a TLS certificate, or any other proven method of authentication. The important thing is that, thanks to the redirect, the control is now out of the hands of the RP and is in complete control of the OP.
```
Comment (justinwb) [inline application-workflow-detailed.md]: typo: it's login screen -> its login screen

File: application-workflow-detailed.md [hunk 14]
```diff
@@ -0,0 +1,442 @@
...
+```
+
+That url might look a little complex, but it's essentially a request to `https://secureauth.example/authorize` with the following url parameters:
+
+ - `scope=open_id`: a list of [oidc scpes](https://auth0.com/docs/scopes/current/oidc-scopes) (ways to communicate what the kind of information the authorization request should wants to get access to). `open_id` is a scope that is needed to verify Alice's identity.
```
Comment (justinwb) [inline application-workflow-detailed.md]: reword this -> "ways to communicate what the kind of information the authorization request should wants to get access to"

**Comment by kjetilk** [comment]
So, I gotta admit that I do not feel quite up to the task of reviewing this for technical accuracy. I think that once the rest of you have worked out the details, I can review it to see if I now understand the documentation, and then, I can merge if I do. OK?

**Comment by zenomt** [comment]
also in the ApplicationWorkflow.png diagram, in the "SENDING REQUEST" section, the steps where Bob's Pod loads Alice's WebID document (to check that Alice's OP is an authorized issuer) and loads Alice's OP's openid-configuration file (so that "5. Requests public keys" can happen), are missing.

**Comment by jaxoncreed** [comment]
@zenomt The request to Alice's Pod was on purpose. Both the Pod and the IDP should provide a well-known/openid-configuration so that clients referencing the Pod can know how to log in even if the IDP is on a separate server.

Though it might be better to do this through a `Link` header. What do you think?

**Comment by jaxoncreed** [comment]
> also in the ApplicationWorkflow.png diagram, in the "SENDING REQUEST" section, the steps where Bob's Pod loads Alice's WebID document (to check that Alice's OP is an authorized issuer) and loads Alice's OP's openid-configuration file (so that "5. Requests public keys" can happen), are missing.

Good point that should be added

**Comment by zenomt** [comment]
> The request to Alice's Pod was on purpose. Both the Pod and the IDP should provide a well-known/openid-configuration so that clients referencing the Pod can know how to log in even if the IDP is on a separate server.

sorry, i don't understand. if the Pod isn't an OpenID Provider, why does it need an openid-configuration? are you sure you didn't mean "if Alice enters her WebID, instead of her OP in step 2, load Alice's WebID Profile to discover her `solid:oidcIssuer`"?  the user interface for logging in directly in your browser is HTML in the `401` response body, and you use a bearer token to access a restricted resource programmatically.

also, i think the main reason for my confusion is that step 2 says "Alice selects her OP", and then step 3 is "request OP configuration from Alice's Pod". that should only happen if Alice selected her Pod as her OP; otherwise RP wouldn't know what Alice's Pod is yet.

regardless, if steps 3 and 4 aren't to Alice's OP, then there also needs to be a transaction before 7 & 8 to load the openid-configuration from Alice's OP.

> Though it might be better to do this through a `Link` header. What do you think?

i'm not a fan of a `Link` header to discover the authorized issuer; please see #18 for a full discussion on that.

File: application-workflow-detailed.md (hunk 7 above)
Comment (jaxoncreed) [reply application-workflow-detailed.md]: For clarification, I changed all references to "app" to "webapp"

**Comment by zenomt** [comment]
> sorry, i don't understand.

the other situation that comes to mind that you might mean is in the case today where solid.community has a wildcard DNS entry to cover all identities *.solid.community, and answers the same openid-configuration file no matter what `Host`/subdomain was requested.

when that's the case, if Alice enters the origin for her webid (which is probably also her pod) instead of her OP URI, RP can ask for `https://alice.solid.community/.well-known/openid-configuration`, and will get back a configuration that's actually for `https://solid.community`.  in particular, the `issuer` claim in the configuration will say `https://solid.community`, not `https://alice.solid.community`.  when that mismatch happens, you need to re-fetch the configuration from the real issuer, to make sure that Alice isn't lying.  otherwise it might be possible for issuer `https://alice.solid.community` to list different endpoints (especially `authorization_endpoint`, `token_endpoint`, and `jwks_uri`) and to issue a counterfeit id_token whose `iss` is `https://solid.community`, which, depending on who received it when, might be believed.

**Comment by jaxoncreed** [comment]
@zenomt You were right the first time. I checked it out and found 2 ways to pick out the issuer from the profile. So I will document them here and remove the requirement for the RS to have a /.well-known/openid-configuration

**Comment by zenomt** [comment]
@jaxoncreed the current revision of ApplicationWorkflow.png AUTHORIZATION step 3 still goes to Alice's Pod, even though the response comes from Alice's OP. i think you just missed that one arc. also 14 "and" should be "an".

to keep the same level of detail as in the rest of the diagram, in "SENDING REQUEST", between box 6 and fetch 7 should be another fetch for the OP configuration by Bob's Pod, so it knows how to request the public keys.  one could argue that "7 requests public keys" means "do the necessary steps to request the public keys", but since it involves 2 round trips, i think it should be called out separately.

**Comment by jaxoncreed** [comment]
@zenomt  right you are

File: application-workflow-detailed.md [hunk 15]
```diff
@@ -467,13 +467,18 @@ A request should be made to url in the `sub` field of the id_token (The user's P
 
 Compare the issuer retrieved in step 5 with the issuer of the id token. If they are not identical the RS must reject the request with a 403.
 
-#### 7. Requests public keys
+#### 7. Retrieves OP Configuration
 
-Now that we've confirmed the validity of the pop_token and the subject of the identity token, we want to ensure the validity of the id_token it contains. To do so, we need the OP's public keys. The OP's address can be obtained via the `iss` field of the id_token (https://secureauth.example). Recall how to retrieve the OP's public keys in steps 3, 4, 7, and 8 of the authorization instructions.
+Before the RS requests information from the thrid-party OP, it must retrieve its configuration. The OP's address can be obtained via the `iss` field of the id_token (https://secureauth.example). Recall how to retrieve the OP's configuration in steps 3 and 4 of the authroization instructions.
```
Comment (TallTed) [inline application-workflow-detailed.md]: typos...
`thrid` -> `third`
`authroization` -> `authorization`

File: application-workflow-detailed.md (hunk 6 above)
Comment (michielbdejong) [reply application-workflow-detailed.md]: can you explain why it is invalid?  I just tried this:
```js
console.log(JSON.parse(`{\
  "grant_types": ["implicit"],\
  "issuer": "https://secureauth.example",\
  "redirect_uris": ["https://www.decentphotos.example/"],\
  "response_types": ["id_token token"],\
  "scope": "openid profile"\
}`))
```
output:
```js
{ grant_types: [ 'implicit' ],
  issuer: 'https://secureauth.example',
  redirect_uris: [ 'https://www.decentphotos.example/' ],
  response_types: [ 'id_token token' ],
  scope: 'openid profile' }
```

File: application-workflow-detailed.md [hunk 16]
```diff
@@ -0,0 +1,494 @@
...
+```
+
+That URL might look a little complex, but it's essentially a request to `https://secureauth.example/authorize` with the following URL parameters:
+
+ - `scope=open_id`: a list of [oidc scpes](https://auth0.com/docs/scopes/current/oidc-scopes) (attributes of the RS to which this token should have access). `open_id` is a scope that is needed to verify Alice's identity.
```
Comment (TallTed) [inline application-workflow-detailed.md]: `[oidc scpes]`
-> `[OIDC scopes]`

File: application-workflow-detailed.md [hunk 17]
```diff
@@ -0,0 +1,494 @@
...
+ - `client_id=7243fd594bdcf9c71a9b902274afaa30`: indicates the id of the client. The value for this field should be obtained in the registration phase.
+ - `response_type=id_token%20token` indicates the desired response data. Note that you cannot use response types that were not previously indicated during registration.
+ - `request=eyJhbGciOiJub25lIn0.eyJyZWRpc...`: A JWT containing the public key of the client and signed by the client using the private key. This is unique to webId-oidc. We will eventually use this to generate our pop-token.
```
Comment (TallTed) [inline application-workflow-detailed.md]: `webId-oidc`
-> `WebID-OIDC`

File: application-workflow-detailed.md [hunk 18]
```diff
@@ -331,10 +331,10 @@ GET https://secureauth.example/authorize?scope=openid&client_id=7243fd594bdcf9c7
...
+ - `scope=open_id`: a list of [OIDC scpes](https://auth0.com/docs/scopes/current/oidc-scopes) (attributes of the RS to which this token should have access). `open_id` is a scope that is needed to verify Alice's identity.
  - `client_id=7243fd594bdcf9c71a9b902274afaa30`: indicates the id of the client. The value for this field should be obtained in the registration phase.
  - `response_type=id_token%20token` indicates the desired response data. Note that you cannot use response types that were not previously indicated during registration.
- - `request=eyJhbGciOiJub25lIn0.eyJyZWRpc...`: A JWT containing the public key of the client and signed by the client using the private key. This is unique to webId-oidc. We will eventually use this to generate our pop-token.
+ - `request=eyJhbGciOiJub25lIn0.eyJyZWRpc...`: A JWT containing the public key of the client and signed by the client using the private key. This is unique to WebId-OIDC. We will eventually use this to generate our pop-token.
```
Comment (TallTed) [inline application-workflow-detailed.md]: hopefully clearer...
`WebId`
-> `WebID`

**Comment by jaxoncreed** [comment]
Can we resurrect this and approve it @kjetilk. I keep sending links to this document because people keep asking how the login works and can't find this.

**Comment by kjetilk** [comment]
> Can we resurrect this and approve it @kjetilk. I keep sending links to this document because people keep asking how the login works and can't find this.

Right! But as of now, it has no approving reviews... As I said above, I'd prefer that others provide reviews and then I review and merge it, because I don't feel up to the task of performing a technical review myself. So, @justinwb , @RubenVerborgh , @TallTed , @michielbdejong , please have another look and update your reviews.

**Comment by RubenVerborgh** [comment]
I can't; not knowledgeable enough about OIDC.

**Comment by kjetilk** [comment]
My problem too... I think the [Authentication Panel](https://github.com/solid/culture/blob/master/panels.md#authentication-panel) should convene urgently to resolve this.

**Comment by dmitrizagidulin** [comment]
@kjetilk 
> I think the Authentication Panel should convene urgently to resolve this.

I agree!

**Review by dmitrizagidulin — DISMISSED** [review]
Read through this carefully on a call with @jaxoncreed.

Looks great, thank you SO much for doing this work, Jackson!

File: application-workflow-detailed.md [hunk 19]
```diff
@@ -331,12 +331,12 @@ GET https://secureauth.example/authorize?scope=openid&client_id=7243fd594bdcf9c7
 
 That URL might look a little complex, but it's essentially a request to `https://secureauth.example/authorize` with the following URL parameters:
 
- - `scope=open_id`: a list of [OIDC scpes](https://auth0.com/docs/scopes/current/oidc-scopes) (attributes of the RS to which this token should have access). `open_id` is a scope that is needed to verify Alice's identity.
+ - `scope=openid`: a list of [OIDC scopes](https://auth0.com/docs/scopes/current/oidc-scopes) (attributes of the RS to which this token should have access). `open_id` is a scope that is needed to verify Alice's identity.
```
Comment (TallTed) [inline application-workflow-detailed.md]:
```
`scope=open_id`
``` 
became
```
`scope=openid`
```
so
```
`open_id` is a scope
```
should become
```
`openid` is a scope
```

**Comment by justinwb** [comment]
Should link to the new document at application-workflow-detailed.md, but the links to this and application-workflow.md should be put in context based on the use case. File names may also need to be changed to align with that and prevent confusion.

**Review by justinwb — APPROVED** [review]
Nice improvement to the current state of the documentation. There's still content required to provide full coverage, especially as it relates to the different patterns of use and how to apply them, but I think that this moves the ball forward in this current instantiation of the spec. The rest of the efforts should be channeled towards https://github.com/solid/specification. With that in mind, I'm good with approving these changes here.

**Review by dmitrizagidulin — APPROVED** [review]

**Pull request merged**